client/src/components/File.js

```diff
@@ -2,225 +2,143 @@
 
 import React from 'react';
 import TransitionGroup from "react-addons-css-transition-group";
-import { getHumanReadableBytes } from '../utils/utils.js';
-import apiurl from 'utils/apiurl';
-import 'styles/File.scss';
-import Highlight from 'components/plugins/highlight';
+import Clipboard from 'clipboard';
 import highlight from 'highlight.js'
+import Highlight from 'components/plugins/highlight';
+import { getHumanReadableBytes } from '../utils/utils.js';
 import ChannelActions from 'actions/ChannelActions';
-import Clipboard from 'clipboard';
-
 import MessageStore from 'stores/MessageStore';
-
+import 'styles/File.scss';
 
 import Logger from 'logplease';
 const logger = Logger.create('Clipboard', { color: Logger.Colors.Magenta });
 
-var getFileUrl = apiurl.getApiUrl() + "/api/cat/";
+function readUTF8String(bytes) {
+  let i = 0;
+  let string = '';
+  if (bytes.slice(0, 3) == '\xEF\xBB\xBF') {
+    i = 3;
+  }
+  for (; i < bytes.byteLength; i++) {
+    const byte1 = bytes[i];
+    if (byte1 < 0x80) {
+      string += String.fromCharCode(byte1);
+    } else if (byte1 >= 0xc2 && byte1 < 0xe0) {
+      const byte2 = bytes[++i];
+      string += String.fromCharCode(((byte1 & 0x1f) << 6) + (byte2 & 0x3f));
+    } else if (byte1 >= 0xe0 && byte1 < 0xf0) {
+      const byte2 = bytes[++i];
+      const byte3 = bytes[++i];
+      string += String.fromCharCode(((byte1 & 0xff) << 12) + ((byte2 & 0x3f) << 6) + (byte3 & 0x3f));
+    } else if (byte1 >= 0xf0 && byte1 < 0xf5) {
+      const byte2 = bytes[++i];
+      const byte3 = bytes[++i];
+      const byte4 = bytes[++i];
+      const codepoint = (((byte1 & 0x07) << 18) + ((byte2 & 0x3f) << 12) + ((byte3 & 0x3f) << 6) + (byte4 & 0x3f)) - 0x10000;
+      string += String.fromCharCode((codepoint >> 10) + 0xd800, (codepoint & 0x3ff) + 0xdc00);
+    }
+  }
+  return string;
+}
 
```

Comment:
Can you elaborate a little on what this block does? Why do we need it and what are the different cases it handles?

As a starting point, let's put the magic hex number to constants that describe what they are. Second, it's not always clear to everyone how bitshifts work, so it'd be good to provide some comments on what's happening at each line where we do bitshifting.